File: src/tools/github_tools.py
import os
import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry
from . import ToolError
from .git_tools import get_current_repo


GITHUB_API_URL = "https://api.github.com"


def _get_session():
    """
    Creates a requests session with a retry strategy for transient server errors.
    """
    session = requests.Session()
    retry_strategy = Retry(
        total=3, # Total number of retries
        backoff_factor=1, # Wait 1s, 2s, 4s between retries
        status_forcelist=[429, 500, 502, 503, 504], # Retry on these status codes
        allowed_methods=["HEAD", "GET", "POST", "PUT", "PATCH", "DELETE"]
    )
    adapter = HTTPAdapter(max_retries=retry_strategy)
    session.mount("https://", adapter)
    session.mount("http://", adapter)
    return session


def _get_headers():
    token = os.getenv("GITHUB_TOKEN")
    if not token:
        raise ToolError("GITHUB_TOKEN no configurado.")
    return {"Authorization": f"token {token}", "Accept": "application/vnd.github.v3+json"}


def _resolve_repo(repo: str) -> str:
    raw = repo if repo else get_current_repo()
    raw = raw.replace("git@github.com:", "").replace("https://github.com/", "").replace(".git", "")
    return raw


def list_issues(repo: str = ""):
    session = _get_session()
    headers = _get_headers()
    target_repo = _resolve_repo(repo)
    url = f"{GITHUB_API_URL}/repos/{target_repo}/issues"
    response = session.get(url, headers=headers)
    response.raise_for_status()
    return response.json()


def create_issue(title: str, body: str, repo: str = ""):
    session = _get_session()
    headers = _get_headers()
    target_repo = _resolve_repo(repo)
    url = f"{GITHUB_API_URL}/repos/{target_repo}/issues"
    response = session.post(url, headers=headers, json={"title": title, "body": body})
    response.raise_for_status()
    data = response.json()
    return f"Issue #{data['number']} created successfully."


def edit_issue(issue_number: int, repo: str = "", title: str = None, body: str = None):
    session = _get_session()
    headers = _get_headers()
    target_repo = _resolve_repo(repo)
    url = f"{GITHUB_API_URL}/repos/{target_repo}/issues/{issue_number}"
    data = {k: v for k, v in {"title": title, "body": body}.items() if v}
    response = session.patch(url, headers=headers, json=data)
    response.raise_for_status()
    return f"Issue #{issue_number} edited successfully."


def close_issue(issue_number: int, repo: str = ""):
    session = _get_session()
    headers = _get_headers()
    target_repo = _resolve_repo(repo)
    url = f"{GITHUB_API_URL}/repos/{target_repo}/issues/{issue_number}"
    response = session.patch(url, headers=headers, json={"state": "closed"})
    response.raise_for_status()
    return f"Issue #{issue_number} closed successfully."


def post_comment(issue_number: int, body: str, repo: str = ""):
    session = _get_session()
    headers = _get_headers()
    target_repo = _resolve_repo(repo)
    url = f"{GITHUB_API_URL}/repos/{target_repo}/issues/{issue_number}/comments"
    response = session.post(url, headers=headers, json={"body": body})
    response.raise_for_status()
    return f"Comment posted to issue #{issue_number}."


def add_label(issue_number: int, label: str, repo: str = ""):
    session = _get_session()
    headers = _get_headers()
    target_repo = _resolve_repo(repo)
    url = f"{GITHUB_API_URL}/repos/{target_repo}/issues/{issue_number}/labels"
    response = session.post(url, headers=headers, json={"labels": [label]})
    response.raise_for_status()
    return f"Label '{label}' added to issue #{issue_number}."


def remove_label(issue_number: int, label: str, repo: str = ""):
    session = _get_session()
    headers = _get_headers()
    target_repo = _resolve_repo(repo)
    url = f"{GITHUB_API_URL}/repos/{target_repo}/issues/{issue_number}/labels/{label}"
    response = session.delete(url, headers=headers)
    response.raise_for_status()
    return f"Label '{label}' removed from issue #{issue_number}."


def get_issue_comments(issue_number: int, repo: str = ""):
    session = _get_session()
    headers = _get_headers()
    target_repo = _resolve_repo(repo)
    url = f"{GITHUB_API_URL}/repos/{target_repo}/issues/{issue_number}/comments"
    response = session.get(url, headers=headers)
    response.raise_for_status()
    return response.json()

File: src/tools/githubTools.js
import axios from "axios";
import axiosRetry from "axios-retry";
import GitTools from "./gitTools.js";

/**
 * Tools for interacting with the GitHub API.
 */
export default class GithubTools {
  #GITHUB_API_URL = "https://api.github.com";

  #client;

  #gitTools;

  /**
   * Initializes the GithubTools instance and configures the axios client with retry logic.
   */
  constructor() {
    this.#client = axios.create({
      baseURL: this.#GITHUB_API_URL,
      headers: {
        Accept: "application/vnd.github.v3+json",
      },
    });

    axiosRetry(this.#client, {
      retries: 3,
      retryDelay: axiosRetry.exponentialDelay,
      /**
       * Determines if the request should be retried based on the error.
       * @param {Error} error - The error object encountered during the request.
       * @returns {boolean} True if the request should be retried.
       */
      retryCondition: (error) =>
        axiosRetry.isNetworkOrIdempotentRequestError(error) ||
        error.response?.status === 429,
    });

    this.#gitTools = new GitTools();
  }

  /**
   * Retrieves the authentication headers using the GITHUB_TOKEN environment variable.
   * @returns {Promise<{Authorization: string}>} The headers object.
   * @throws {Error} If GITHUB_TOKEN is not configured.
   */
  async #getHeaders() {
    const token = process.env.GITHUB_TOKEN;
    if (!token) {
    throw new Error("GITHUB_TOKEN no configurado.");
  }
    return { Authorization: `token ${token}` };
  }

  /**
   * Resolves a repository identifier to a standard "owner/repo" format.
   * @param {string} repo The repository URL or identifier.
   * @returns {string} The resolved repository name.
   */
  #resolveRepo(repo) {
    const raw = repo || "";
    return raw
      .replace("git@github.com:", "")
      .replace("https://github.com/", "")
      .replace(".git", "");
  }

  /**
   * Prepares the repository identifier and authentication headers for a request.
   * @param {string} [repo] The repository identifier.
   * @returns {Promise<{targetRepo: string, headers: object}>} The prepared request data.
   */
  async #prepareRequest(repo) {
    const targetRepo = repo
      ? this.#resolveRepo(repo)
      : await this.#gitTools.getCurrentRepo();

    if (!targetRepo) {
      throw new Error("Could not resolve target repository.");
    }

    const headers = await this.#getHeaders();
    return { targetRepo, headers };
  }

  /**
   * Gets the details of the currently authenticated user.
   * @returns {Promise<object>} The user data.
   */
  async getAuthenticatedUser() {
    const headers = await this.#getHeaders();
    const response = await this.#client.get("/user", { headers });
    return response.data;
  }

  /**
   * Lists issues for a given repository.
   * @param {string} [repo] The repository identifier.
   * @returns {Promise<Array>} A list of issues.
   */
  async listIssues(repo) {
    const { targetRepo, headers } = await this.#prepareRequest(repo);
    const response = await this.#client.get(`/repos/${targetRepo}/issues`, { headers });
    return response.data;
  }

  /**
   * Creates a new issue in a repository.
   * @param {string} title The title of the issue.
   * @param {string} body The body text of the issue.
   * @param {string} [repo] The repository identifier.
   * @returns {Promise<string>} A success message.
   */
  async createIssue(title, body, repo) {
    const { targetRepo, headers } = await this.#prepareRequest(repo);
    const response = await this.#client.post(
      `/repos/${targetRepo}/issues`,
      { title, body },
      { headers },
    );
    return `Issue #${response.data.number} created successfully.`;
  }

  /**
   * Edits an existing issue.
   * @param {number|string} issueNumber The issue number to edit.
   * @param {string} [repo] The repository identifier.
   * @param {string} [title] The new title for the issue.
   * @param {string} [body] The new body for the issue.
   * @returns {Promise<string>} A success message.
   */
  async editIssue(issueNumber, repo, title, body) {
    const { targetRepo, headers } = await this.#prepareRequest(repo);
    const data = {};
    if (title) {
    data.title = title;
  }
    if (body) {
    data.body = body;
  }

    await this.#client.patch(
      `/repos/${targetRepo}/issues/${issueNumber}`,
      data,
      { headers },
    );
    return `Issue #${issueNumber} edited successfully.`;
  }

  /**
   * Closes an existing issue.
   * @param {number|string} issueNumber The issue number to close.
   * @param {string} [repo] The repository identifier.
   * @returns {Promise<string>} A success message.
   */
  async closeIssue(issueNumber, repo) {
    const { targetRepo, headers } = await this.#prepareRequest(repo);
    await this.#client.patch(
      `/repos/${targetRepo}/issues/${issueNumber}`,
      { state: "closed" },
      { headers },
    );
    return `Issue #${issueNumber} closed successfully.`;
  }

  /**
   * Posts a comment to an issue.
   * @param {number|string} issueNumber The issue number to comment on.
   * @param {string} body The content of the comment.
   * @param {string} [repo] The repository identifier.
   * @returns {Promise<string>} A success message.
   */
  async postComment(issueNumber, body, repo) {
    const { targetRepo, headers } = await this.#prepareRequest(repo);

    const cleanIssueNumber = String(issueNumber).replace(/\D/g, '');
    if (!cleanIssueNumber) {
    throw new Error("Invalid issue number provided.");
  }

    if (!body || typeof body !== 'string' || body.trim() === '') {
      throw new Error("Comment body must be a non-empty string.");
    }

    await this.#client.post(
      `/repos/${targetRepo}/issues/${cleanIssueNumber}/comments`,
      { body: body.trim() },
      { headers },
    );
    return `Comment posted to issue #${cleanIssueNumber}.`;
  }

  /**
   * Adds a label to an issue.
   * @param {number|string} issueNumber The issue number to label.
   * @param {string} label The label to add.
   * @param {string} [repo] The repository identifier.
   * @returns {Promise<string>} A success message.
   */
  async addLabel(issueNumber, label, repo) {
    const { targetRepo, headers } = await this.#prepareRequest(repo);
    await this.#client.post(
      `/repos/${targetRepo}/issues/${issueNumber}/labels`,
      { labels: [label] },
      { headers },
    );
    return `Label '${label}' added to issue #${issueNumber}.`;
  }

  /**
   * Removes a label from an issue.
   * @param {number|string} issueNumber The issue number to modify.
   * @param {string} label The label to remove.
   * @param {string} [repo] The repository identifier.
   * @returns {Promise<string>} A success message.
   */
  async removeLabel(issueNumber, label, repo) {
    const { targetRepo, headers } = await this.#prepareRequest(repo);
    await this.#client.delete(
      `/repos/${targetRepo}/issues/${issueNumber}/labels/${label}`,
      { headers },
    );
    return `Label '${label}' removed from issue #${issueNumber}.`;
  }

  /**
   * Retrieves comments for a specific issue.
   * @param {number|string} issueNumber The issue number.
   * @param {string} [repo] The repository identifier.
   * @returns {Promise<Array>} A list of comments.
   */
  async getIssueComments(issueNumber, repo) {
    const { targetRepo, headers } = await this.#prepareRequest(repo);
    const response = await this.#client.get(
      `/repos/${targetRepo}/issues/${issueNumber}/comments`,
      { headers },
    );
    return response.data;
  }
}